docs/madr/decisions/010-timeout-policy.md

```diff
@@ -0,0 +1,174 @@
+Ad 2. Simpler config and you can still achieve the same result as previously by configuring it for `MeshSubset|MeshService|MeshServiceSubset`
+
+#### Connection and stream timeout
+I understand that `max connection timeout` and `max stream timeout ` are basically doing the same thing, since every 
+connection event in HTTP 1 are treated as streams in Envoy according to the docs.
+
+#### Route timeouts
+As for the route timeouts, we only care about `idle timeout` and `timeout` since `per try timeout` and `per try idle timeout` 
+are retry related timeouts and should be configured in retry policy.
+
+#### Request headers timeout
+Last one specific timeout is `request headers timeout` but this timeout makes no sense on outbound connections. 
+This only can be used to protect yourself from malicious inbound clients.
+
+#### Scaled timeouts
+Moreover, there is a mechanism of scaled timeout in envoy, which I am not sure if we want to expose to our users, at least not at first.
+
+### Specification
+Top-level targetRef can have all available kinds:
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+  name: ...
+```
+
+MeshUpstreamTimeout is outbound policy, that's why it has a "to" array:
+```yaml
+to:
+  - targetRef:
+      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshHTTPRoute
+      name: ...
+```
+
+#### Minimal configuration
+Taking into account previous considerations, the minimal configuration for `MeshUpstreamTimeout` policy should look like this:
+```yaml
+default:
+  connectTimeout: ... # all protocols
+  idleTimeout: ... # all protocols
+  http:
+    requestTimeout: ... 
+    streamIdleTimeout: ...
+    maxStreamDuration: ...
+```
+
+
+### Examples
+#### Default configuration
+```yaml
+type: MeshUpstreamTimeout  
+mesh: default  
+name: default-timeouts  
+spec:  
+  targetRef:  
+    kind: Mesh  
+    name: default  
+  to:  
+    - targetRef:  
+        kind: Mesh  
+        name: default  
+      default:  
+        connectTimeout: 5s  
+        idleTimeout: 1h  
+        http:  
+          requestTimeout: 15s  
+          streamIdleTimeout: 30m  
+          maxStreamDuration: 0s
+```
+
+#### Configuration with overridden TCP specific timeouts for database
+
+```yaml
+type: MeshUpstreamTimeout  
+mesh: default  
+name: default-timeouts  
+spec:  
+  targetRef:  
+    kind: Mesh  
+    name: default  
+  to:  
+    - targetRef:  
+        kind: Mesh  
+        name: default  
+      default:  
+        connectTimeout: 5s  
+        idleTimeout: 1h  
+        http:  
+          requestTimeout: 15s  
+          streamIdleTimeout: 30m  
+          maxStreamDuration: 0s  
+---  
+type: MeshUpstreamTimeout  
+mesh: default  
+name: database-tcp-timeouts  
+spec:  
+  targetRef:  
+    kind: Mesh  
+    name: default  
+  to:  
+    - targetRef:  
+        kind: MeshService  
+        name: postgres  
+      default:  
+        connectTimeout: 1s  
+        idleTimeout: 10m
+```
```

Comment:
I think we need some special handling for routes. Let's say we have this `default-timeouts` in example above, but we also define this

```
type: MeshUpstreamTimeout  
mesh: default  
name: route-tcp-timeouts  
spec:  
  targetRef:  
    kind: Mesh  
    name: default  
  to:  
    - targetRef:  
        kind: MeshHTTPRoute  
        name: route-xyz
      default:  
        connectTimeout: 1s  
        idleTimeout: 10m
        http:
          requestTimeout: 15s  
          streamIdleTimeout: 30m  
          maxStreamDuration: 0s
```

I know that we don't have `MeshHTTPRoute` at this very moment, but still.
If we have such a situation, `connectTimeout` & `idleTimeout` are defined in the Cluster which is shared between many routes. Same with `streamIdleTimeout` which is configured on HTTPConnectionManager which is shared between many routes.

Should we just introduce validation that if you are using targetRef MeshHTTPRoute/MeshGatewayRoute you cannot use those values?

Reply:
That's a very good point indeed.

Reply:
@Automaat please take note of this validation in MADR

Reply:
this validation is described in [To level](https://github.com/kumahq/kuma/pull/5094/files#diff-0253c70bc803371f20ff4699e1368086d76285337e7465d8ce40a4237bddfb10R109) section: 
```
For now this policy will be implemented only for `Mesh` and `MeshService` in the future support for `MeshHTTPRoute` can be added.
For `MeshHTTPRoute` there are few limitations. According to docs for route we can only configure: `idleTimeout`,
`requestTimeout` and `maxStreamDuration`. In the future when adding support for `MeshHTTPRoute` we can validate if only allowed timeouts are used.
```

Will it be enough, or should I add this as a separate section?

Reply:
It's good enough